plugins/contents/fps_contents/routes.py

```diff
@@ -186,6 +186,10 @@ async def read_content(
                         if "metadata" not in cell:
                             cell["metadata"] = {}
                         cell["metadata"].update({"trusted": False})
+                        if cell["cell_type"] == "code":
+                            cell_source = cell["source"]
+                            if not isinstance(cell_source, str):
+                                cell["source"] = "".join(cell_source)
```

Comment:
Any reason `jupyverse` does not use `nbformat` like in Jupyter Server?

https://github.com/jupyter-server/jupyter_server/blob/3544deb53902cc02c9aa9d6513b3c30f1113896b/jupyter_server/services/contents/fileio.py#L264-L291

Since reading with `nbformat` should ensure the `source` will be a single string: https://nbformat.readthedocs.io/en/latest/format_description.html#cell-types

Reply:
Jupyverse generally reboots the whole Jupyter server stack, it doesn't use jupyter-client either for instance. An exception is nbconvert though, so I'm not against revisiting the use of nbformat, but I'm not totally convinced either.
This is not the only place where nbformat is not used, there is also [jupyter-ydoc](https://github.com/jupyter-server/jupyter_ydoc/issues/46).

Reply:
ok. It's just that `nbformat` also contains some validation which can be useful in Jupyverse. Also it is unlikely to change that much over time as the notebook format has been quite stable.